xcm/xcm-builder/src/pay.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! `PayOverXcm` struct for paying through XCM and getting the status back
+
+use frame_support::traits::{
+	tokens::{Pay, PaymentStatus},
+	Get,
+};
+use sp_std::{marker::PhantomData, vec};
+use xcm::prelude::*;
+use xcm_executor::traits::{QueryResponseStatus, XcmQueryHandler};
+
+// `PayOverXcm` is a type which implements the `frame_support_traits::tokens::Pay` trait, to allow
+// for generic payments of a given `AssetKind` and `Balance` from an implied target, to a
+// beneficiary via XCM, and relying on an XCM `TransferAsset` operation.
+//
+// `PayOverXcm::pay` is asynchronous, and returns a `QueryId` which can then be used via
+// `check_payment` to check the status of the xcm transaction transaction.
+//
```

Comment:
Should use doc comments, not simple code comments:
```suggestion
///
```